src/mlpack/base.hpp

```diff
@@ -0,0 +1,111 @@
+ */
+#ifndef MLPACK_BASE_HPP
+#define MLPACK_BASE_HPP
+
+// First, check if Armadillo was included before, warning if so.
+#ifdef ARMA_INCLUDES
+#pragma message "Armadillo was included before mlpack; this can sometimes cause\
+ problems.  It should only be necessary to include <mlpack/core.hpp> and not \
+<armadillo>."
+#endif
+
+// Defining _USE_MATH_DEFINES should set M_PI.
+#define _USE_MATH_DEFINES
+#include <cmath>
+
+// Next, standard includes.
+#include <cctype>
+#include <cfloat>
+#include <climits>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+#include <numeric>
+#include <vector>
+#include <queue>
+
+// But if it's not defined, we'll do it.
+#ifndef M_PI
+  #define M_PI 3.141592653589793238462643383279
+#endif
+
+// MLPACK_COUT_STREAM is used to change the default stream for printing
+// purpose.
+#if !defined(MLPACK_COUT_STREAM)
+ #define MLPACK_COUT_STREAM std::cout
+#endif
+
+// MLPACK_CERR_STREAM is used to change the stream for printing warnings
+// and errors.
+#if !defined(MLPACK_CERR_STREAM)
+ #define MLPACK_CERR_STREAM std::cerr
+#endif
+
+// Give ourselves a nice way to force functions to be inline if we need.
+#undef mlpack_force_inline
+#define mlpack_force_inline
+#if defined(__GNUG__) && !defined(DEBUG)
+  #undef mlpack_force_inline
+  #define mlpack_force_inline __attribute__((always_inline))
+#elif defined(_MSC_VER) && !defined(DEBUG)
+  #undef mlpack_force_inline
+  #define mlpack_force_inline __forceinline
+#endif
+
+// Backport std::any from C+17 to C++11 to replace boost::any.
+// Use mnmlstc backport implementation only if compiler does not
+// support C++17.
+#if __cplusplus < 201703L
+  #include <mlpack/core/std_backport/any.hpp>
+  #include <mlpack/core/std_backport/string_view.hpp>
+  #define MLPACK_ANY core::v2::any
+  #define MLPACK_ANY_CAST core::v2::any_cast
+  #define MLPACK_STRING_VIEW core::v2::string_view
+#else
+  #include <any>
+  #include <string_view>
+  #define MLPACK_ANY std::any
+  #define MLPACK_ANY_CAST std::any_cast
+  #define MLPACK_STRING_VIEW std::string_view
+#endif
+
+// Now include Armadillo through the special mlpack extensions.
+#include <mlpack/core/arma_extend/arma_extend.hpp>
+#include <mlpack/core/util/arma_traits.hpp>
+
+// On Visual Studio, disable C4519 (default arguments for function templates)
+// since it's by default an error, which doesn't even make any sense because
+// it's part of the C++11 standard.
+#ifdef _MSC_VER
+  #pragma warning(disable : 4519)
+#endif
+
+// Ensure that the user isn't doing something stupid with their Armadillo
+// defines.
+#include <mlpack/core/util/arma_config_check.hpp>
+
+// This can be removed with Visual Studio supports an OpenMP version with
+// unsigned loop variables.
+#if (defined(_OPENMP) && (_OPENMP >= 201107))
+  #undef  MLPACK_USE_OPENMP
+  #define MLPACK_USE_OPENMP
+#endif
+
```

Comment:
It seems that elsewhere in the mlpack codease,  uses of  `#pragma omp parallel for` are not guarded by `#if defined(MLPACK_USE_OPENMP) ... #endif`.  This may cause compile-time warnings about unused `#pragmas` if OpenMP is not enabled, or will blow up on "compilers" that don't support OpenMP 3.1+.

Suggest to add the following to address this:
```
#if defined(MLPACK_USE_OPENMP)
  #define MLPACK_OMP_FOR _Pragma("omp parallel for")
#else
  #define MLPACK_OMP_FOR
#endif
```
The `MLPACK_OMP_FOR` macro can then be used as a safe replacement for all unguarded instances of `#pragma omp parallel for`.

Explanation of the _Pragma() directive: https://en.cppreference.com/w/cpp/preprocessor/impl

Reply:
Do you have any examples of compilers where compilation will actually entirely fail?  I'm not convinced that the extra overhead burden of `MLPACK_OMP_FOR` is worth it.  I don't mind if users get warnings when OpenMP isn't enabled---in my view, there is a benefit to using the "standard" OpenMP syntax; someone coming across the code already familiar with OpenMP (which is probably a sizeable subset of C++ programmers) doesn't need to wonder what is going on.